lib/src/internal/http/http_route.dart

```diff
@@ -0,0 +1,116 @@
+import 'http_route_param.dart';
+import 'http_route_part.dart';
+
+/// Builds routes according to Discord's dynamic bucket rate limiting scheme.
+///
+/// Use builder syntax such as:
+/// ```dart
+/// var route = HttpRoute()..guilds(id: id)..members()..search();
+/// ```
+/// to keep route definitions brief while reusing route rate limiting definitions.
+/// If creating custom routes with [add], remember to comply with Discord's
+/// rate limiting scheme by toggling the appropriate [HttpRouteParam.isMajor].
+class HttpRoute {
+  final List<HttpRoutePart> _httpRouteParts = [];
+
+  List<String> get pathSegments => _httpRouteParts
+      .expand((part) => [
+            part.path,
+            ...part.params.map((param) => param.param),
+          ])
+      .toList();
+
+  String get path => "/" + pathSegments.join("/");
+
+  String get routeId => _httpRouteParts
+      .expand((part) => [
+            part.path,
+            ...List.generate(
+              part.params.length,
+              (index) => part.params[index].isMajor ? part.params[index].param : r"$param",
```

Comment:
why `r"$param"` here?

Reply:
So that non-major parameters all get encoded as `$param` (literally) in the id string, meaning they don't influence the id of the request ratelimit bucket. The value could be absolutely anything but `$param` makes it clear for debugging that it would be substituted with the value of a minor parameter